src/Core/Managed/Shared/Extensibility/MetricAggregator.cs

```diff
@@ -0,0 +1,172 @@
+        public override int GetHashCode()
+        {
+            return this.hashCode;
+        }
+
+        /// <summary>
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="other">The object to compare with the current object. </param>
+        /// <returns>True if the specified object is equal to the current object; otherwise, false.</returns>
+        public bool Equals(MetricAggregator other)
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            return string.Compare(this.aggregatorId, other.aggregatorId, StringComparison.Ordinal) == 0;
+        }
+
+        /// <summary>
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object. </param>
+        /// <returns>True if the specified object is equal to the current object; otherwise, false.</returns>
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj))
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, obj))
+            {
+                return true;
+            }
+
+            return this.Equals(obj as MetricAggregator);
+        }
+
+        /// <summary>
+        /// Generates id of the aggregator serving time series specified in the parameters.
+        /// </summary>
+        /// <param name="metricName">Metric name.</param>
+        /// <param name="dimensions">Optional metric dimensions.</param>
+        /// <returns>Aggregator id that can be used to get aggregator.</returns>
+        private static string GetAggregatorId(string metricName, IDictionary<string, string> dimensions = null)
+        {
+            StringBuilder aggregatorIdBuilder = new StringBuilder(metricName ?? string.Empty);
+
+            if (dimensions != null)
+            {
+                var sortedDimensions = dimensions.OrderBy((pair) => { return pair.Key; });
+
+                foreach (KeyValuePair<string, string> pair in sortedDimensions)
+                {
+                    aggregatorIdBuilder.AppendFormat(CultureInfo.InvariantCulture, "\n{0}\t{1}", pair.Key ?? string.Empty, pair.Value ?? string.Empty);
+                }
+            }
+
+            return aggregatorIdBuilder.ToString();
+        }
+
+        /// <summary>
+        /// Forwards value to metric processors.
+        /// </summary>
+        /// <param name="value">Value tracked on time series.</param>
+        private void ForwardToProcessors(double value)
+        {
+            // create a local reference to metric processor collection
+            // if collection changes after that - it will be copied not affecting local reference
+            IList<IMetricProcessor> metricProcessors = this.manager.Client.TelemetryConfiguration.MetricProcessors;
```

Comment:
Ouch. This is a pretty bad violation of the [law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter). From customer viewpoint, this means that in order to test the use of `MetricAggregator` in my code, I'll need to understand a chain of objects 4 levels deep and create it for every test. In practice, I think few people will go through the effort successfully and will decide to wrap the `MetricAggregator` and the `MetricAggregatorManager` in application abstractions as "untestable external dependencies" instead. Is it possible to make `IEnumerable<IMetricProcessor>` a direct dependency of the `MetricAggregator`?